stellargraph/layer/graph_classification.py

```diff
@@ -0,0 +1,236 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.keras import backend as K
+from tensorflow.keras import activations, initializers, constraints, regularizers
+from tensorflow.keras.layers import (
+    Input,
+    Layer,
+    Lambda,
+    Dropout,
+    Reshape,
+    Masking,
+    GlobalAveragePooling1D,
+)
+
+from ..mapper import GraphGenerator
+from .cluster_gcn import ClusterGraphConvolution
+from ..core.experimental import experimental
+
+
+@experimental(reason="Missing unit tests")
+class GraphClassificationConvolution(ClusterGraphConvolution):
+
+    """
+    A graph convolutional Keras layer. A stack of such layers can be used to create a model for supervised graph
+    classification.
+
+    The implementation is based on the GCN Keras layer of keras-gcn github
+    repo https://github.com/tkipf/keras-gcn
+
+    Notes:
+      - The inputs are tensors with a batch dimension.
+
+      - There are 2 inputs required, the node features and the normalized graph adjacency matrices for each batch entry.
+
+      - This class assumes that the normalized graph adjacency matrices are passed as input to the Keras methods.
+
+    Args:
+        units (int): dimensionality of output feature vectors
+        activation (str): nonlinear activation applied to layer's output to obtain output features
+        use_bias (bool): toggles an optional bias
+        kernel_initializer (str): name of layer bias f the initializer for kernel parameters (weights)
+        bias_initializer (str): name of the initializer for bias
+        kernel_regularizer (str): name of regularizer to be applied to layer kernel. Must be a Keras regularizer.
+        bias_regularizer (str): name of regularizer to be applied to layer bias. Must be a Keras regularizer.
+        activity_regularizer (str): not used in the current implementation
+        kernel_constraint (str): constraint applied to layer's kernel
+        bias_constraint (str): constraint applied to layer's bias
+    """
+
+    def call(self, inputs):
+        """
+        Applies the layer.
+
+        Args:
+            inputs (list or tuple): a list or tuple of 2 input tensors that includes
+                node features (size batch_size x N x F),
+                graph adjacency matrix (batch_size x N x N),
+                where N is the number of nodes in the graph, and F is the dimensionality of node features.
+
+        Returns:
+            Keras Tensor that represents the output of the layer.
+        """
+        features, A = inputs
+
+        h_graph = K.batch_dot(A, features)
+        output = K.dot(h_graph, self.kernel)
+
+        # Add optional bias & apply activation
+        if self.bias is not None:
+            output += self.bias
+        output = self.activation(output)
+
+        return output
+
+
+@experimental(reason="Missing unit tests")
+class GraphClassification:
```

Comment:
This seems to be a very general name for what seems to be a specific solution to graph classification (although clearly one that has some flexibility), could you talk a bit about why this is the right name?

Reply:
Essentially, a general name for a very generic algorithm. Most other models we implement already have names as published, e.g., GCN, GAT, GraphWave, DDGK, DGCNN, etc. but this one is so generic that it is hard to come up with a good name; any suggestions are welcome.

That said, I will rename it `SupervisedGraphClassification` because it is more specific.

Reply:
I don't want to argue too much about what to name things, but as I said before many models and layers have distinct names in the literature so those will be easy to name. This model, however, does not so we can name it whatever we want. I don't particularly like very long names so I'll just suggest we go with `SupervisedGCNGraphClassification`, that is drop mention of node pooling in the name and just have it explained in the documentation. This way, if we want to use a different pooling layer, we won't have to rename the model class as well. I had a preference for just `SupervisedGraphClassification` because then we could have extended both the model and the pooling method without renaming the model class but happy to go with the more specific `SupervisedGCNGraphClassification` or maybe `GCNSupervisedGraphClassification`. Just let me know which name your prefer of all the ones suggested in this thread.

Reply:
I'm fine either `SupervisedGCNGraphClassification` or `GCNSupervisedGraphClassification `, with a mild preference for the latter, because it potentially makes the connection to a more general `SupervisedGraphClassification` class clearer 👍 



> I had a preference for just SupervisedGraphClassification because then we could have extended both the model and the pooling method without renaming the model class 

My concern is that we won't be able to do this in a great way. To do an extension with the class directly, we'll likely need to add even more arguments to have all the options for all of the supported algorithms (e.g. `attn_heads` and `attn_heads_reduction` for GAT, `teleport_probability` for APPNP) as well as some custom way for the user to tell the graph classification code which to use, even though we've already got most of the pieces of that: the individual model classes (each takes the appropriate arguments, and "which to use" is the user choosing the class they want).

I thus suspect we may want a noticeably different API for a general form, such as something more similar to `DeepGraphInfomax` where we can reuse those existing model-defining classes (this may require some careful adjustments to those classes too):

```python
class SupervisedGraphClassification:
    def __init__(self, base_model, generator, pooling="mean"):
        ...

    ...
```

The existing `GCNSupervisedGraphClassification` can then be implemented in terms of it, without us having to do any complicated argument overloading tricks:

```python
class GCNSupervisedGraphClassification(SupervisedGraphClassification):
    def __init__(self, layer_sizes, activations, generator, ..., pooling="mean"):
        base_model = GCN(layer_sizes, activations, ...)
        super().__init__(self, base_model, generator, pooling)
```

(We can then deprecate this specialised class if we wish, or leave it as a convenience. Also, this is all predicated on GCN/GAT/... working for graph classification.)

Do you have a different idea for how you were planning to generalise, that I'm not understanding?

> This way, if we want to use a different pooling layer, we won't have to rename the model class as well

Yeah, I guess it's easier to be able to swap a single layer simple layer like this in or out (as my hypothetical `pooling` argument above demonstrates), and so there's not so many subtle future generalisation/migration concerns. 👍

Reply:
I hope we can agree that for now, we can keep it as is and after we have implemented a couple more graph classification algorithms, we can revisit and re-factor as necessary. We should probably discuss bigger picture design issues either in a different ticket or offline.

Reply:
I was explaining my reasoning to make sure we were on the same page since your mention of extending this class made me wonder if you had some other plan in mind that allowed us to have the good name right now. In addition, I was careful to the code examples tiny to just cover the core idea and only explore the design space necessary to guide the names in PR.

(Also, since there was no comment like "I went with `GCNSupervisedGraphClassification`", I didn't realise that you'd decided upon, edited and pushed a new name.)